pam_gitdap.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PAM_GITDAP" "8" "August 2020" "" ""
.
.SH "NAME"
\fBpam_gitdap\fR \- PAM module for authenticating git users against LDAP
.
.SH "SYNOPSIS"
\fI[service\-name]\fR auth \fIcontrol\-flag\fR pam_gitdap \fI\fImodule configuration\fR\fR
.
.SH "DESCRIPTION"
\fBpam_gitdap\fR provides PAM authentication through a simple LDAP bind\. It is designed to allow a user\'s LDAP username/password to be used as a second authentication factor for git hosting services such as Gitea or Gogs\.
.
.SH "MODULE CONFIGURATION"
The following values are mandatory in the PAM module configuration:
.
.TP
\fBldap_uri=\fR\fIURI\fR
The LDAP URI address for the LDAP server to bind to\. Format is \fIldap://host:port\fR or \fIldaps://host:port\fR
.
.TP
\fBdn_attr=\fR\fIAttribute\fR
The Bind DN naming attribute, for example \fIcn\fR or \fIuid\fR\.
.
.TP
\fBdn_base=\fR\fIBase\fR
The base of the Bind DN, appended after the naming attribute\.
.
.br
For example \fIou=users,dc=example,dc=com\fR\.
.
.P
The following values are optional in the PAM configuration:
.
.TP
\fBprompt_user\fR
Prompt for the git username\. This is used if multiple git users share a single ssh account, which is typically the case for services like Gitea\.
.
.TP
\fBdebug\fR
Enables module debugging\.
.
.SH "USAGE RECOMMENDATIONS"
It is recommended to create a second SSH server exclusively for git, running on an alternate port\. You may do this by creating a symbolic link from your system \fBsshd\fR binary to an alternate name such as \fB/usr/local/sbin/git\-sshd\fR\.
.
.P
Create a separate PAM configuration for the \fBgit\-sshd\fR service which includes an \fBauth\fR entry for this module along with any other PAM modules you wish to use\.
.
.P
The \fIsshd_config\fR file for this server must have the \fBChallengeResponseAuthentication\fR option set to \fByes\fR to allow the PAM module to prompt for the LDAP password and/or username\.
.
.br
It should also have \fBAllowUsers\fR set to only your git service user and \fBAuthenticationMethods\fR set to \fBpublickey,keyboard\-interactive\fR\. This will require your users to prevent a valid SSH key before being prompted for their LDAP credentials\.
.
.P
\fBIT IS NOT RECOMMENDED TO USE THIS MODULE AS YOUR ONLY AUTHENTICATION METHOD\. IT SHOULD ALWAYS BE USED IN CONJUNCTION WITH SSH KEYS AS DESCRIBED ABOVE, OR AS PART OF A MORE COMPREHENSIVE PAM STACK\.\fR
.
.SH "MODULE OPERATION"
The module will attempt to bind to the LDAP directory as
.
.br
"\fBdn_attr\fR=\fIusername\fR,\fBdn_base\fR" using the password supplied by the user\.
.
.P
It returns one of the following PAM result codes:
.
.TP
\fBPAM_SUCCESS\fR
The LDAP bind succeeded\.
.
.TP
\fBPAM_AUTH_ERR\fR
The LDAP bind failed\.
.
.TP
\fBPAM_SERVICE_ERR\fR
The module configuration is invalid (one or more mandatory values missing)\.
.
.TP
\fBPAM_BUF_ERR\fR
An internal error occurred in string processing/manipulation\.
.
.SH "BUGS"
\fBOnly a simple bind is performed\fR
.
.br
This is insecure unless used over TLS (ldaps://) connections\.
.
.P
\fBNo validation is performed on user\-supplied usernames\fR This is intentional as it allows for flexibility in the username\.
.
.br
For example if configured with \fBdn_attr= dn_base= prompt_user\fR the username expected would be the full LDAP DN to bind as\.
.
.P
\fBNo validation is performed on dn_attr or dn_base\fR
.
.br
This is intentional as it allows more complex Bind DNs to be constructed\.
.
.SH "COPYRIGHT"
pam_gitdap is copyright (C) 2020 Michael Graziano\.
.
.br
Source is available at \fIhttps://github\.com/voretaq7/pam_gitdap\fR
.
.SH "SEE ALSO"
pam(8), pam\.conf(5)


pam_gitdap.c
/* "pam_gitdap" - an LDAP authentication module for PAM
 * (C) 2020 - Michael Graziano (mikeg@bsd-box.net)
 *
 * This module is intended to be used in a git (gitea)
 * SSH server instance as a second factor authenticating
 * git users against an external LDAP directory.
 * 
 * This module is intended to be used as a second authentication
 * layer in addition to SSH keys.
 *
 * PAM Parameters
 *   ldap_uri=		The LDAP server URI
 *   dn_attr=		The DN naming attribute for users
 *   dn_base=		Where to look for the named users
 *   prompt_user	Force a username prompt
 *			(DOES NOT change pam_user value in the stack)
 *   debug		Turns on debug mode
 *
 * e.g. with ldap_uri=ldaps://127.0.0.1 , dn_attr=cn,
 *           dn_base=ou=users,dc=example,dc=com
 * the module will attempt to bind to LDAP Server at 127.0.0.1 (with SSL)
 * as cn=<Username>,ou=users,dc=example,dc=com with the password provided.
 *
 * If the bind succeeds the user is authorized ; if it fails the user
 * is NOT authorized.
 *
*/
#include <sys/cdefs.h>

#include <sys/types.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define PAM_SM_AUTH

#include <security/pam_appl.h>
#include <security/pam_modules.h>
#include <security/pam_mod_misc.h>

#include <ldap.h>

PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh, int flags __unused,
    int argc, const char *argv[])
{

	/* General Module Variables */
	int i;
	int retval; 			// Used for internal return values
	int pam_retval = PAM_AUTH_ERR;  // What we actually return
	int style;			// PAM prompt style
	const char *pam_user;		// Used if we're not prompting.
	char *git_user = NULL;		// Used when we are prompting.
	char *git_password = NULL;	// Self Explanatory

	LDAP *ldp;			// LDAP Object Pointer
	char *bind_dn = NULL;		// LDAP Bind DN
	struct berval cred;		// LDAP Bind Password
	struct berval *servcred;	// Just a throwaway...

	int ldap_protocol=LDAP_VERSION3;
	char prompt_user = 0;
	char debug = 0;
	char *ldap_uri = NULL;	// LDAP Server URI
	char *dn_attr = NULL;	// DN naming attribute (like "cn" or "uid")
	char *dn_base = NULL;	// like "ou=users,dc=example,dc=com"

	// Initialize LDAP from Args
	for (i=0 ; i < argc ; i++) {
		if (strncmp(argv[i], "ldap_uri=", 9) == 0) {
			if (asprintf(&ldap_uri, "%s", argv[i] + 9) == -1) {
				pam_retval = PAM_BUF_ERR;
				goto PAM_GITDAP_CLEANUP;
			}
			continue;
		}
		if (strncmp(argv[i], "dn_attr=", 8) == 0) {
			if (asprintf(&dn_attr, "%s", argv[i] + 8) == -1) {
				pam_retval = PAM_BUF_ERR;
				goto PAM_GITDAP_CLEANUP;
			}
			continue;
		}
		if (strncmp(argv[i], "dn_base=", 8) == 0) {
			if (asprintf(&dn_base, "%s", argv[i] + 8) == -1) {
				pam_retval = PAM_BUF_ERR;
				goto PAM_GITDAP_CLEANUP;
			}
			continue;
		}
		if (strncmp(argv[i], "ldap_proto=", 11) == 0) {
			ldap_protocol = atoi(argv[i] + 11);
			if ( (ldap_protocol > LDAP_VERSION_MAX) ||
			     (ldap_protocol < LDAP_VERSION_MIN) ) {
				PAM_LOG("pam_gitdap Bad Protocol: Using v3.");
				ldap_protocol = LDAP_VERSION3;
			}
			continue;
		}
		if (strncmp(argv[i], "prompt_user", 11) == 0) {
			prompt_user=1;
			continue;
		}
		if (strncmp(argv[i], "debug", 5) == 0) {
			debug=1;
			continue;
		}
	}

	// Validate PAM parameters
	if ( (ldap_uri == NULL) || (dn_base == NULL) || (dn_attr == NULL) ) {
		PAM_LOG("pam_gitdap ldap_uri, dn_base, or dn_attr missing.");
		PAM_LOG("           Check PAM configuration.");
		pam_retval = PAM_SERVICE_ERR;
		goto PAM_GITDAP_CLEANUP;
	}

	// Get username
	if (prompt_user) {
		style = PAM_PROMPT_ECHO_ON;
		retval = pam_prompt(pamh, style, &git_user, "git User: ");
		if (retval != PAM_SUCCESS) {
			pam_retval = retval;
			goto PAM_GITDAP_CLEANUP;
		}
	} else {
		// Use the PAM user
		retval = pam_get_user(pamh, &pam_user, NULL);
		if (retval != PAM_SUCCESS) {
			pam_retval = retval;
			goto PAM_GITDAP_CLEANUP;
		} else {
			if (asprintf(&git_user, "%s", pam_user) == -1) {
				pam_retval = PAM_BUF_ERR;
				goto PAM_GITDAP_CLEANUP;
			}
		}
	}

	/* Get Password
	 *
	 * It doesn't make sense to use a password that has already been
	 * typed in since we haven't presented the challenge to the user
	 * yet, so clear the stored password.
	 */
	pam_set_item(pamh, PAM_AUTHTOK, NULL);
	style = PAM_PROMPT_ECHO_OFF;
	retval = pam_prompt(pamh, style, &git_password, "git Password: ");
	if (retval != PAM_SUCCESS) {
		pam_retval = PAM_BUF_ERR;
		goto PAM_GITDAP_CLEANUP;
	}
	pam_set_item(pamh, PAM_AUTHTOK, git_password);

	// Initialize LDAP credential (Password)
	cred.bv_val = git_password;
	cred.bv_len = strlen(git_password);

	// Initialize LDAP username (dn)
	if (asprintf(&bind_dn, "%s=%s,%s", dn_attr,
		      git_user, dn_base) == -1) {
		pam_retval = PAM_BUF_ERR;
		goto PAM_GITDAP_CLEANUP;
	}
	PAM_LOG("pam_gitdap BindDN: %s", bind_dn);
	PAM_LOG("pam_gitdap Server: %s", ldap_uri);

	/*
	 * Actually connect to the LDAP server and try to bind.
	*/
	if (ldap_initialize(&ldp, ldap_uri) == LDAP_SUCCESS) {
		if ( ldap_set_option(ldp, LDAP_OPT_PROTOCOL_VERSION,
		      &ldap_protocol) == LDAP_SUCCESS ) {
			retval = ldap_sasl_bind_s(ldp, bind_dn,
			         LDAP_SASL_SIMPLE, &cred,
			         NULL, NULL, &servcred);
			PAM_LOG("LDAP Bind: %s", ldap_err2string(retval));
			if (retval == LDAP_SUCCESS) {
				pam_retval = PAM_SUCCESS;
				ldap_unbind_ext_s(ldp, NULL, NULL);
			} else {
				pam_retval = PAM_AUTH_ERR;
			}
		} else {
			PAM_LOG("LDAP Couldn't set LDAP protocol %d",
				 ldap_protocol);
			pam_retval = PAM_SERVICE_ERR;
		}
	} else {
		PAM_LOG("pam_gitdap Initialize failed: %s",
			ldap_err2string(retval));
		pam_retval = PAM_SERVICE_ERR;
	}


	// Clean up everything we threw on the heap
	PAM_GITDAP_CLEANUP:	// Shut up I know GOTOs are bad.
	free(git_user);
	free(git_password);
	free(bind_dn);
	free(ldap_uri);
	free(dn_attr);
	free(dn_base);

	// Return the PAM status
	return (pam_retval);
}

PAM_EXTERN int
pam_sm_setcred(pam_handle_t *pamh __unused, int flags __unused,
    int argc __unused, const char *argv[] __unused)
{

	return (PAM_SUCCESS);
}

PAM_MODULE_ENTRY("pam_gitdap");
